File: Code/PS2/Matlab/ccprofit.m
function [profit] = ccprofit(nfirms,descn,binom,D,f,ggamma)
% For a given number of active firms, computes the profit and market share 
% for the static Cournot competition for each firm in all  states. 

% initialize function output
profit = zeros(descn,nfirms); % Profit for each state-firm

for i = 1:descn
    if mod(i, 50) == 0 % report progress
        fprintf('  Computed: %d\n', i);
    end

    w = decode(i,nfirms,binom); % decode state i to state w
    theta = ggamma * exp(-(w-4));  % marginal cost

    %   Solve for cournot equilibrium with n firms; reduce n until all firms
    %   want to produce quantity > 0
    n=nfirms;
    p = (D + sum(theta(1:n)))/(n+1); % eqm price for n firms
    while ~((p - theta(n) >= 0) || (n==1)) % reduce n until either positive q or 1 firm left 
        n=n-1;
        p = (D + sum(theta(1:n)))/(n+1);
    end
    q = zeros(nfirms,1);
    if p - theta(n) > 0
        q(1:n) = p - theta(1:n); % eqm quantity
    end
    quan=q;

    pstar = D - sum(quan);   % Equilibrium price
    profstar = (pstar>theta).*(pstar-theta).*quan - f; % Equilibrium profits

    profit(i,:) = profstar'; 
end

File: Code/PS2/Python/ccprofit.py
def ccprofit(nfirms, descn, binom, D, f, ggamma):
    """
    Computes the profit and market share for the static Cournot competition 
    for each firm in all states.

    Args:
        nfirms (int): Number of active firms.
        descn (int): Number of possible industry structures.
        binom (numpy.ndarray): Binomial coefficient matrix.
        D (float): Cournot demand intercept.
        f (float): Fixed cost per firm.
        ggamma (float): Marginal cost coefficient.

    Returns:
        numpy.ndarray: Profit matrix of shape (descn, nfirms).
    """
    profit = np.zeros((descn, nfirms))  # Initialize profit matrix

    for i in range(descn):
        if i % 50 == 0:  # Print progress every 50 iterations
            print(f"  Computed: {i}")

        w = decode(i, nfirms, binom)  # Decode state i into efficiency levels
        theta = ggamma * np.exp(-(np.array(w) - 4))  # Compute marginal cost

        # Solve for Cournot equilibrium: Reduce number of firms until all produce positive quantity
        n = nfirms
        p = (D + np.sum(theta[:n])) / (n + 1)  # Equilibrium price

        while not ((p - theta[n - 1] >= 0) or (n == 1)):  # Reduce n if price makes last firm unprofitable
            n -= 1
            p = (D + np.sum(theta[:n])) / (n + 1)

        q = np.zeros(nfirms)  # Initialize output quantities
        if p - theta[n - 1] > 0:  # Ensure positive quantity production
            q[:n] = p - theta[:n]

        quan = q  # Equilibrium quantity

        pstar = D - np.sum(quan)  # Equilibrium price
        profstar = (pstar > theta) * (pstar - theta) * quan - f  # Compute profits

        profit[i, :] = profstar  # Store computed profits

    return profit
